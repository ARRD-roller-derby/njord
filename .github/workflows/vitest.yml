name: Vitest
on: push
jobs:
  build:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2
    - uses: actions/setup-node@v2
      with:
        node-version: '16.x'
    - name: Install dependencies
      run: npm ci
    - name: Run Vitest tests
      run: npm run coverage
    - uses: actions/upload-artifact@v2
      if: always()
      with:
        name: vitest-report
        path: vitest-report/
        retention-days: 30