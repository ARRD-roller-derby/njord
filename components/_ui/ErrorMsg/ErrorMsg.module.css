.container {
  text-align: center;
  color:var(--color-txt-error);
  font-size: 1.1em;
}