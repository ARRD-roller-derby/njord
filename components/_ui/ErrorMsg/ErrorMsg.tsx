import classes from './ErrorMsg.module.css';

interface props {
  readonly message: string
}
export default function ErrorMsg({message}:props){

  return <div className={classes.container}>{message}</div>
}