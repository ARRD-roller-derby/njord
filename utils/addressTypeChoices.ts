import { addressType } from "../types/addressType.enum";

export const addressTypeChoices = [
  {
    label: 'Mon adresse',
    value:addressType.personnal
  },
  {
    label: 'Stade',
    value:addressType.stadium
  },
  {
    label: 'Point de covoiturage',
    value:addressType.carpool
  },
  {
    label: 'Autre',
    value:addressType.other
  }
]