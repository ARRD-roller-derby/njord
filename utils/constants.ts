export const DISCORD_NEWS_HOOK = process.env.DISCORD_NEWS_HOOK || ''