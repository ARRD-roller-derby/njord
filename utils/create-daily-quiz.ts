import dayjs from "dayjs"
import Question from "../models/question.model"
import Quiz from "../models/quiz.model"
import { QuizType } from "../types/quiz.interface"
import { percent } from "./percent"
import { questionDifficulty } from "./question-difficulty"
import { shuffle } from "./shuffle"
import { MongoDb } from '../db/mongo.connect'

export async function createDailyQuiz() {

  await MongoDb()

  const existDailyQuiz = await Quiz.findOne({
    type: QuizType.daily,
    day: dayjs().format('YYYY-MM-DD')
  })

  if (existDailyQuiz) return existDailyQuiz

  // Search old Quizzes
  const oldsQuizzes = await Quiz.find({
    type: QuizType.daily, day: {
      $in: [
        dayjs().subtract(4, 'day').format('YYYY-MM-DD'),
        dayjs().subtract(3, 'day').format('YYYY-MM-DD'),
        dayjs().subtract(2, 'day').format('YYYY-MM-DD'),
        dayjs().subtract(1, 'day').format('YYYY-MM-DD')
      ]
    },
    answser: { $exists: true, $not: { $size: 0 } },
  }).select('questions')

  //Take olds questions
  const oldsQuestions = oldsQuizzes.reduce((current: string[], state) => {
    current.push(...state.questions)
    return current
  }, [])

  //Take new questions without olds questions
  const questions = await Question.find({ active: true, _id: { $not: { $in: oldsQuestions } } }).select('_id bad_answers_num good_answers_num')
  const shuffleQuestions = shuffle<string>(questions.map(question => question._id), 4)

  const difficulties = questions
    .filter((question) => shuffleQuestions.includes(question._id))

  //For difficulty
  const questionsNum = difficulties.reduce((current: { bad: number, good: number }, state) => {
    current.bad += state.bad_answers_num
    current.good += state.good_answers_num
    return current
  }, { bad: 0, good: 0 })

  return await Quiz.create({
    difficulty: questionDifficulty(
      percent(
        questionsNum.good,
        questionsNum.good + questionsNum.bad
      )
    ),
    questions: shuffleQuestions,
    type: QuizType.daily,
    day: dayjs().format('YYYY-MM-DD')
  })

} 