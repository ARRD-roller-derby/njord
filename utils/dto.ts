import validator from 'validator'

function sanitizeString(value: string): string | boolean {
  return  validator.escape(value || '')
}

export default function dto<T>(dto: Object, obj: T): undefined | T {
  let count = 0
  const objKeys = Object.keys(dto),
    newObj = { ...obj }

  //delete unused key
  Object.keys(obj).forEach((key) => {
    if (!dto.hasOwnProperty(key)) {
      delete newObj[key]
    }
  })

  objKeys.forEach((key) => {
    //False if one or more required key is not in obj
    if (dto[key].hasOwnProperty('isRequired')) {
      if (!newObj.hasOwnProperty(key)) {
        count++
      }
    }

    //exemple email: 'string'
    if (typeof dto[key] === 'string') {
      if (dto[key] === 'array') {
        if (!Array.isArray(newObj[key])) {
          count++
        }
      } else if (dto[key] !== typeof newObj[key] && newObj[key]) {
        count++
      }

      if (dto[key] === 'string' && dto[key] !== undefined ) {
        newObj[key] = sanitizeString(newObj[key])
      }
    }
    // exemple email : {type:'string',isRequired:true}
    else {
      if (!dto[key].hasOwnProperty('type')) {
        count++
      }
      if (dto[key].type !== typeof newObj[key]) {
        count++
      }

      if (dto[key].type === 'string') {
        newObj[key] = sanitizeString(newObj[key])
      }
    }
  })
  return count ? undefined : newObj
}
