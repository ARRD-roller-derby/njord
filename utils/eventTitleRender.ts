import { EventInterface } from '../types/Event.interface'
import { EventType } from '../types/EventType.enum'

export default function eventTitleRender(event: EventInterface):string {
  if (event.title) return event.title

  if (event.type === EventType.training) return 'entraînement'
  if (event.type === EventType.generalAssembly) return 'AG'
  if (event.type === EventType.other) return 'autre'
  if (event.type === EventType.online) return 'en ligne'

  return 'événement'
}
