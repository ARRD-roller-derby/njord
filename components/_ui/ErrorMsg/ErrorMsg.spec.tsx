import {describe,expect, it} from 'vitest'
import { cleanup } from '@testing-library/react'
import ErrorMsg from './ErrorMsg'
import { render, screen } from '../../../utils/test-utils'

describe('<ErrorMsg />', () => {
  afterEach(cleanup)

  it('Check snapshot', async () => {
    const {asFragment } = render(
        <ErrorMsg message='erreur'/>
    )

    expect(await screen.findByText('erreur')).toBeInTheDocument()
    expect(asFragment()).toMatchSnapshot()
  })
})
