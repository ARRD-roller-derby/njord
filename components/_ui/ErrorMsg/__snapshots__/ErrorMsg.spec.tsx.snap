// Vitest Snapshot v1

exports[`<ErrorMsg /> > Check snapshot 1`] = `
<DocumentFragment>
  <div
    class="_container_sm64p_1"
  >
    erreur
  </div>
</DocumentFragment>
`;
